### ISSUE TITLE
Possible Implementation errors for first tournament strategies

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#1273]
@id428 has written a detailed pdf (attached to this issue)
[First Tournament Strategies.pdf](https://github.com/Axelrod-Python/Axelrod/files/3841326/First.Tournament.Strategies.pdf)
 with their cross examination of the strategies which is attached to the issue. Here is a summary they posted on gitter:

> In summary, these are the errors we found:
> - Tideman & Chieruzzi: The strategy reportedly defected in the last two rounds. This is not implemented. (There are also two minor additional points mentioned in the PDF)
> - Nydegger: The value 49 is missing in A
> - Grofmann: We dont know where the behaviour for the first seven rounds is described. Maybe something was implemented that was not existend in the first place.
> - Shubik: This strategy should cooperate at least twice after each run of retaliations. The implementation prevents that. (At least that's how we understand Axelrods description, see PDF)
> - Graaskamp: Reportedly checks for its own twin but does not in the implementation
> - Downing: We don't know where the final decision rule in the implementation comes from.

**Comment by marcharper** [comment]
(1) I think we should put a big disclaimer on this file that says that the first tournament source code doesn't exist and some of the strategies are not sufficiently and unambiguously described to ensure that they match the original tournament. These are not necessarily implementation "errors" as much as "choices".

(2) Since there have been other errors in Axelrod's original works (like the number of rounds that each tournament was run, and bugs we found in the second tournament's source code) we can't assume that the descriptions in any given text (or the second tournament code's comments) are "correct".

(3) Tideman & Chieruzzi: it's not clear what "defected in the last two rounds" means. Was that by chance or was the strategy intended to always defect on the last two rounds? In the second tournament players did not know how many rounds the game would run, only the current round number.  T&C is in the second tournament with the following [source](http://www-personal.umich.edu/%7Eaxe/research/Software/CC/CC2/TourExec1.1.f.html):

      FUNCTION K84R(JP,M,IS,JS,R, JA)
C BY T NICOLAUS TIDEMAN AND PAULA CHIERUZZI
C TYPED BY JM 3/31/79
    k84r=ja    ! Added 7/27/93 to report own old value
      K84R = 1
      IF (M .GT. 1) GOTO 2
      ISIG = 0
      DS = 0
      JQ = 0
      FJD = 0
      JDR = 0
      FM = 0
      GOTO 3
2     IF (JP .EQ. 1) FJD = FJD + 1
      IF (ISIG .EQ. 1) GOTO 5
      FM = M
      IF (JQ .EQ. 0 .AND. JP .EQ. 1) JDR = JDR + 1
3     IF (IS - JS - DS - 5 * JDR * (JDR - 1) / 2 .GE. 0)
     1  K84R = 0
      IF (K84R .EQ. 1) GOTO 4
1     JQ = JP
      RETURN
4     IF ((JQ - JP) .LT. 0 .OR. (M - ISIG) .LT. 10) GOTO 1
      IF (ABS(FJD - (FM - 1.) / 2.) .LT. (1.5 * SQRT(FM - 1.)
     1 )) GOTO 1
      ISIG = 1
      JQ = JP
      GOTO 6
5     ISIG = M
      JQ = 0
      JDR = 0
      DS = IS - JS
6     K84R = 0
      RETURN
      END
      FUNCTION K32R(J,M,K,L,R, JA)

I don't see any obvious check to always defect on the last 2 rounds (maybe I'm just not seeing it?). The code may have also been changed from the first tournament version.

(4) Nydegger missing 49: this one seems like an implementation error in the Python version (49 appears in the second tournament code)

C BY RUDY NYDEGGER
C TYPED BY AX, 3/27/79 (SAME AS ROUND ONE NYDEGR)
c Replaced by Nydegr retyped from rnd 1 by Ax, 7/27/93
c   T=0
c    K57R=NYDEGR(J,M,K,L,T,R)
c    RETURN
c   END
    k57r=ja    ! Added 7/27/93 to report own old value
    if(m.ne.1) goto 5
    k57r = 0
    n = 0
c update 3 move history
    5   N = 4 * (n-16*(N/16)) + 2 * k57r + J
    if(m.gt.3) goto 8
    k57r=j
    if(m.eq.3 .and. n.eq.6) k57r=1
    return
c coop if 0, 27, 28, 32, 40-4, 46-8, 56-7,59-60,62-63
   8    k57r=1
    if(n-39) 10,110,50
 10 if(n) 100,100,20
 20 if(n-28) 30,100,40
 30 if(n-27) 110,100,100
 40 if(n-32) 110,100,110
 50 if(n-45) 100,110,60
 60 if(n-49) 100,110,70
 70 if(n-58) 80,110,90
 80 if(n-55) 110,110,100
 90 if(n-61) 100,110,100
 100    k57r = 0
 110    return
    end

(5) Grofman behaves differently on rounds before 7 according to the second tournament source code:

C BY BERNARD GROFMAN
C FROM CARDS BY JM 3/27/79
        DIMENSION IOPPNT(999)
    k86r=ja    ! Added 7/27/93 to report own old value
        IOPPNT(MOVEN) = JPICK
        MYOLD = K86R
        IF (MOVEN .GT. 2) GOTO 10
        K86R = 0
        RETURN
10      IF (MOVEN. GT. 7) GOTO 20
        K86R = JPICK
        RETURN
20      IPREV7 = 0
      J = MOVEN - 7
      K = MOVEN - 1
        DO 25 I = J,K
25              IPREV7 = IPREV7 + IOPPNT(I)
        IF (MYOLD .EQ. 0 .AND. IPREV7 .LE. 2) K86R = 0
        IF (MYOLD .EQ. 0 .AND. IPREV7 .GT. 2) K86R = 1
        IF (MYOLD .EQ. 1 .AND. IPREV7 .LE. 1) K86R = 0
        IF (MYOLD .EQ. 1 .AND. IPREV7 .GT. 1) K86R = 1
        RETURN
    END

(6) Shubik -- I'm not sure about this one, the descriptions are hard to parse / somewhat ambiguous IMO. It's not in the second tournament source.

(7) Graaskamp round (1) is in the second tournament source code and claims to match the first round source. There was an abandoned attempt to convert the Fortran to python in [PR 1115](https://github.com/Axelrod-Python/Axelrod/pull/1115). The python source clearly states it does not match the original, so again not an error as much as a choice of implementation based on the description.

 Integer FUNCTION GRASR(JPICK, MOVEN, ISCOR, JSCOR, RANDO,JA)
    DIMENSION NMOV(4)
    GRASR=JA        ! Added 7/32/93 to report own old value
c Next line for debugging
c   if(moven. eq. 57)  write(6,99) jscor
c99 format(' TEST from GRASR at move 57. jscor = ', i6)
      IF (MOVEN .NE. 1) GO TO 9997
    DO 9996 I = 1, 4
    NMOV(I) = 0
9996    CONTINUE
    NMOVE = 0
    IGAME = 0
    N = 0
9997    CONTINUE
    IF (MOVEN - 1) 25, 25, 26
25  GRASR = 0
    RETURN
26  IF (MOVEN - 51) 1, 2, 3
1   GRASR = JPICK
    RETURN
2   GRASR = 1
    RETURN
3   IF (MOVEN - 57) 4, 5, 6
4   IF (MOVEN - 52) 9, 9, 10
10  NMOV(MOVEN - 52) = MMOVE + JPICK
9   GRASR = JPICK
    IF (GRASR -1) 7, 8, 8
7   MMOVE = 2
    GO TO 11
8   MMOVE = 4
11  RETURN
5   IF (JSCOR - 135) 19, 19, 20
20  J = NMOV(2)
    GO TO (12, 12, 30, 31, 32), J
31  IF (NMOV(1) - 3) 12, 35, 12
35  IF (NMOV(3) - 3) 12, 16, 12
32  IF (NMOV(1) - 5) 12, 33, 12
33  IF (NMOV(3) - 5) 12, 16, 12
30  IF (NMOV(1) - 2) 12, 34, 12
34  IF (NMOV(3) - 4) 12, 40, 12
40  IF (NMOV(4) - 2) 12, 41, 12
12  IGAME = 1
    N = RANDO * 10.0 + 5.0
    GRASR = 0
    RETURN
16   IGAME = 2
    GRASR = 0
    RETURN
19   IGAME = 3
27   GRASR = 1
    RETURN
41   IGAME = 4
42   GRASR = 0
    IF (MOVEN - 118) 44, 43, 43
43  IGAME=2
44  RETURN
6   GO TO (21, 22, 27, 42), IGAME
21  IF (N) 23, 23, 24
23  GRASR = 1
    N = RANDO * 10.0 + 5.0
    RETURN
24  GRASR = JPICK
    N = N-1
    RETURN
22  GRASR = JPICK
    RETURN
    END

(8) AFAIK only the code for Revised Downing is available. Somewhere in the literature it is mentioned that Downing had a bug (hence the revised version). The code for Revised Downing is in the second tournament and is the source of the decision rule:

      FUNCTION K59R(LASTMV,MOVEN,K,L,R, JA)
C BY LESLIE DOWNING
C TYPED BY AX, 3/27/79 (SAME AS ROUND ONE REV.DOWNING)
c Redone as copy of K56=RevDowning by Ax, 7/27/93
c     INTEGER XDOWNC
c      T=0
c      K59R=XDOWNC(J,M,K,L,T,R)
c     RETURN
c     END
       INTEGER LASTMV,MOVEN
      INTEGER PAST,NICE1,NICE2
      REAL GOOD,BAD,C,ALT
      INTEGER DEFECT, COOP
      INTEGER TOTCOP,TOTDEF
    k59r=ja    ! Added 7/27/93 to report own old value
      IF (MOVEN - 2) 1,4,2
1     GOOD = 1.0
      BAD = 0.0
      PAST = 0
      TOTCOP = 0
      TOTDEF = 0
      NICE1 = 0
      NICE2 = 0
      COOP = 0
      DEFECT = 1
      GOTO 4
2     IF (PAST .EQ. DEFECT) GOTO 3
      IF (LASTMV .EQ. COOP) NICE1 = NICE1 + 1
      TOTCOP = TOTCOP + 1
      GOOD = FLOAT(NICE1) / FLOAT(TOTCOP)
      GOTO 4
3     IF (LASTMV .EQ. COOP) NICE2 = NICE2 + 1
      TOTDEF = TOTDEF + 1
      BAD = FLOAT(NICE2) / FLOAT(TOTDEF)
4     PAST = K59R
      C = 6.0 * GOOD - 8.0 * BAD - 2.0
      ALT = 4.0 * GOOD - 5.0 * BAD - 1.0
      IF (C .GE. 0.0 .AND. C .GE. ALT) GOTO 5
      IF (C .GE. 0.0 .AND. C .LT. ALT) GOTO 6
      IF (ALT .GE. 0.0) GOTO 6
      K59R = DEFECT
      GOTO 7
5     K59R = COOP
      GOTO 7
6     K59R = 1 - K59R
7     RETURN
      END

**Comment by marcharper** [comment]
A meta comment on the second round source -- many of the strategies explicitly say that they match the "round one" version, and I'm not sure if we should assume so if such a comment is not present. However I'm inclined to do so since having some source code reference is better than not having it at all.

**Comment by id428** [comment]
You got some very good points there. We were aware of the source code for the second tournament but we know literally nothing about Fortran and understood nothing of it, which is why we decided to instead focus on Axelrods Paper and its references/citations. 
Take what you deem valuable from our comments. We just thought you might have overlooked some things could use these remarks. If you have good reasons to do what we could not reconstruct, that's great!

**Comment by marcharper** [comment]
Thanks for looking through the code. I think we could indicate where we've used the code from the second tournament as a guide, and we can fix the missing 49 for sure.

Unfortunately otherwise, unless some source code or very accurate descriptions surface, we'll never really know how some of the strategies in the first tournament were implemented or if the descriptions accurately match the actual implementations. But going forward we can prevent that from happening again by having openly accessible and well-tested implementations in this library for strategies that can be reproduced.

**Comment by drvinceknight** [comment]
> (1) I think we should put a big disclaimer on this file that says that the first tournament source code doesn't exist and some of the strategies are not sufficiently and unambiguously described to ensure that they match the original tournament. These are not necessarily implementation "errors" as much as "choices".

Yup good idea :+1:

---

Before taking each strategy point by point I think some of the confusion comes from the following general problem:

So far in the implementations of the strategies from Axelrod's two tournaments, a strategy has been named after it's author. We should allow for a differentiation between strategies from the first and second when a difference is known (like the source code of the second **clearly** not matching the description of the first). I would propose:

```
<Author>
```

and 

```
More<Author>
```

for the case of a given author having strategies in both the first and the second tournament. We could also choose to have `More` as a prefix to all strategies from Axelrod's second tournament regardless of whether or not the authors also submitted a strategy to the first.

A good example of this is `Grofman` where the description in the first paper is actually pretty simple but it's currently encoded as the strategy in the second tournament. I would suggest the current implementation be renamed `MoreGrofman` and we implement the strategy from the first from the description as `Grofman`. For some reason I actually thought we had done that but perhaps I'm imagining things.

(My suggestion of `More` as a prefix is based on the second paper being called ”More Effective Choice in the Prisoner's Dilemma" but I'm open to any other suggestions.)

> 
> (3) Tideman & Chieruzzi: it's not clear what "defected in the last two rounds" means. Was that by chance or was the strategy intended to always defect on the last two rounds? In the second tournament players did not know how many rounds the game would run, only the current round number. T&C is in the second tournament with the following [source](http://www-personal.umich.edu/%7Eaxe/research/Software/CC/CC2/TourExec1.1.f.html):

I think the description in the first paper (page 22 of https://www.jstor.org/stable/pdf/173932.pdf?refreqid=excelsior%3A927ed81f07724eec0e1fb65cdaa3cab6) does clearly state that it's meant to defect in the last two round.

So MoreTidemanAndChieruzzi` (which currently doesn't exist) should indeed be based on the Fortran but [TidemanAndChieruzzi](https://axelrod.readthedocs.io/en/stable/_modules/axelrod/strategies/axelrod_first.html#TidemanAndChieruzzi) should have those last two defections and match the description in the paper as closely as possible.

> (4) Nydegger missing 49: this one seems like an implementation error in the Python version (49 appears in the second tournament code)

Yup, we should fix that :+1:

> (5) Grofman behaves differently on rounds before 7 according to the second tournament source code:

This should also be a case for `MoreGrofman` and `Grofman`, the description in the paper is actually pretty clear here (page 22 of https://www.jstor.org/stable/pdf/173932.pdf?refreqid=excelsior%3A927ed81f07724eec0e1fb65cdaa3cab6)

> (6) Shubik -- I'm not sure about this one, the descriptions are hard to parse / somewhat ambiguous IMO. It's not in the second tournament source.

I agree with @marcharper 

> (7) Graaskamp round (1) is in the second tournament source code and claims to match the first round source. There was an abandoned attempt to convert the Fortran to python in [PR 1115](https://github.com/Axelrod-Python/Axelrod/pull/1115). The python source clearly states it does not match the original, so again not an error as much as a choice of implementation based on the description.

I do think we have missed recognising the "twin" though, we have the following description in the docstring:

```
The strategy also checks to see if the opponent is playing Tit For Tat or another strategy from a preliminary tournament called ‘Analogy’ If so it plays Tit For Tat. If not it cooperates and randomly defects every 5 to 15 moves.
```

but the description in the paper states:

```
is TIT FOR TAT, ANALOGY (a program from the preliminary tournament), and its own twin
```

We can't do anything about Analogy but we should check for the twin.

> AFAIK only the code for Revised Downing is available. Somewhere in the literature it is mentioned that Downing had a bug (hence the revised version). The code for Revised Downing is in the second tournament and is the source of the decision rule:

I think we can implement `Downing` from the description in the paper, it estimates the opponent as a reactive strategy and chooses the best option against that.

---

What do you think @marcharper ?

Thanks @id428 for going through this.

**Comment by drvinceknight** [comment]
It looks like at some stage we had some of these implemented correctly, here's a very early `Groffman`: https://github.com/Axelrod-Python/Axelrod/blob/e1d8407724a8449c07d5978919aa3e953ba1f74b/axelrod/strategies/axelrod_first.py#L177

**Comment by marcharper** [comment]
I'm not sure we can trust the published descriptions. Without code we don't have a alternative, but I'd defer to the second tournament code when available even if there are small differences with the published description.

> My suggestion of More as a prefix is based on the second paper being called ”More Effective Choice in the Prisoner's Dilemma" but I'm open to any other suggestions.

How about `GrofmanFirst` and `GrofmanSecond` or `Grofman` and `GrofmanV2`?

> We can't do anything about Analogy but we should check for the twin.

IMO we should translate the Fortran and just have one version. It's not obvious to me how it's doing these checks, but we have the code nevertheless and it says is the same as the code in the first tournament. I'm not sure how we could justify any other heuristic.

> I think we can implement Downing from the description in the paper, it estimates the opponent as a reactive strategy and chooses the best option against that.

Downing takes a [parameter](https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/strategies/axelrod_first.py#L95) `revised` -- maybe we should split into separate strategies?

> I think the description in the first paper [of T&C] does clearly state that it's meant to defect in the last two rounds.

I think this is kind of a weird artifact of the first tournament rules, specifically that the number rounds was fixed. In the second tournament, players don't know how many rounds there are. So is the rule "defect on the last two rounds (if the number of rounds is known)" or "defect on rounds 199 and 200"? Maybe the former is the more likely intention, but the latter is more likely how it was coded, assuming the strategy functions take the same form as in the second tournament. Which should we do?

I'll note again that in the second tournament's implementation the strategy doesn't seem to defect on the last two rounds, which means we're saying that we believe it's different than the first strategy's version (trusting the text description instead) because there's *not* a comment saying that it's the same implementation as in the first tournament.

**Comment by drvinceknight** [comment]
> I'm not sure we can trust the published descriptions. Without code we don't have a alternative, but I'd defer to the second tournament code when available even if there are small differences with the published description.

I think we should consider the first tournament and second tournament as different sets of strategies unless specifically instructed not to. I would suggest that the authors are actually quite likely to have modified their strategy given what they learnt from the first tournament. We should implement the first tournament strategies using the descriptions but highlighting that we're making assumptions and that the descriptions are not clear but of course if we can look at the Fortran and gain some intuition about any potential assumptions then we should do that.

I think what we currently have is a bit different to this though, as the coded first tournament strategies no longer match the descriptions in the paper. I think confusion got introduced with the second tournament strategies "overwriting" the first tournament ones by the same authors.

> How about `GrofmanFirst` and `GrofmanSecond` or `Grofman` and `GrofmanV2`?

I like `First`/ `Second` more than ``/`V2`. Having it as a prefix (so `FirstGrofman`) would allow for automcpletion from `First...` or `Second...` to give us all the strategies in those two tournaments which could be helpful no?

> IMO we should translate the Fortran and just have one version. It's not obvious to me how it's doing these checks, but we have the code nevertheless and it says is the same as the code in the first tournament. I'm not sure how we could justify any other heuristic.

If we implement the Fortran (which we should certainly continue to try an do) then that gives us `SecondGraaskamp` and we cannot claim it to be `FirstGraaskamp`, I think it's ok to implement `FirstGraaskamp` from the description and just be super explicit in the docstring. Note that this would not be far of the status quo: we just omitted the twin check. One further thing we could do is to name them differently, so `SecondGraaskamp` based on the fortran and `AlmostGraaskamp` (which is what we currently have) or something like that?

> Downing takes a [parameter](https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/strategies/axelrod_first.py#L95) `revised` -- maybe we should split into separate strategies?
> 
> >

Sure, I don't fully recall where the `revised` came from, I have some vague recollection that it was not the difference between the first and second tournament but I could be wrong? 


> I think this is kind of a weird artifact of the first tournament rules, specifically that the number rounds was fixed. In the second tournament, players don't know how many rounds there are. So is the rule "defect on the last two rounds (if the number of rounds is known)" or "defect on rounds 199 and 200"? Maybe the former is the more likely intention, but the latter is more likely how it was coded, assuming the strategy functions take the same form as in the second tournament. Which should we do?

So the paper describes it as: 

>  The program defects automatically on the last two moves.

I would suggest we make it defect on game length - 2 (so it's sticking to that wording) and have that correctly classified (note that we probably had this correct previously as the classifier for the strategy currently shows that it uses game length which it does actually not - I suspect we modified the code following the Fortran but didn't change the classifier...). 

In term of "if the number of rounds is known" or not, the library will take care of that anyway, we have other strategies that do similar things. Note that we have the ability to pass through what it known to tournaments/matches: https://axelrod.readthedocs.io/en/stable/tutorials/advanced/player_information.html. So we can play fixed length matches but without the strategies knowing what the length is etc. So for strategies like this those last two defections just never kick in (because they never know they're in those turns).

> I'll note again that in the second tournament's implementation the strategy doesn't seem to defect on the last two rounds, which means we're saying that we believe it's different than the first strategy's version (trusting the text description instead) because there's not a comment saying that it's the same implementation as in the first tournament.

I agree and I believe we should in general assume that the strategies are different unless specifically told otherwise. So I think `FirstGroffman`  and `SecondGroffman` is a good naming convention to go for. I think we should go for `Second<NAME>` even if `<NAME>` did not have an entry in the first tournament. 

(Again I'm not precious about the naming convention, if we think a suffix is better I can be convinced, perhaps being super explicit is good though, maybe something like `FirstTounamentSubmissionByGroffman` or `FirstByGroffman` might not be terrible as it helps clarify the difference between strategy and strategy author?)

**Comment by marcharper** [comment]
I agree that it may be likely that some authors changed their strategies but I also find the descriptions suspect. Some of the strategies have very complicated sequences of conditionals that are simply not well captured in a few lines of prose. My guess is that tweaked code is closer to the original than a description (but that's ultimately just an opinion).

Perhaps we should agree upon rules of precedence for the available information sources. IIUC your proposal is that we prefer information in this order:

1. Use the second tournament code for second tournament players
2. Use the first tournament descriptions for first tournament players over second tournament code, even if the second tournament code says it's the same as the first tournament?) <-- I may be misunderstanding the comments above re: Graaskamp

I propose we meet in the middle at:

1. Use the second tournament code for second tournament players
2. If the 2nd tournament code says a strategy is unchanged, use that for the first tournament also (e.g. for Graaskamp)
3. Otherwise use the first tournament description to make a "best guess" at a reasonable implementation, drawing inspiration from the second tournament code (if any exists)

WDYT?

**Comment by drvinceknight** [comment]
> I propose we meet in the middle at:
> 
> 1. Use the second tournament code for second tournament players
> 2. If the 2nd tournament code says a strategy is unchanged, use that for the first tournament also (e.g. for Graaskamp)
> 3. Otherwise use the first tournament description to make a "best guess" at a reasonable implementation, drawing inspiration from the second tournament code (if any exists)
> 
> WDYT?

This sounds like a perfect way forward! :+1:

**Comment by drvinceknight** [comment]
Do you have a preference about the naming convention?

**Comment by marcharper** [comment]
> How about GrofmanFirst and GrofmanSecond or Grofman and GrofmanV2?

Either is fine with me. I guess I slightly prefer Grofman and GrofmanV2 since in some cases there may only be a v1, but I don't feel strongly about it. In any case we can add a list of `axelrod_first_tournament_strategies` along side the short run time strategies and other lists.

**Comment by drvinceknight** [comment]
Something corresponding to "Version" is probably my least favourite option as I think it makes it sound (very subjectively) like they could be slight tweaks (which in practice they could be of course but are not necessarily).I'd prefer something corresponding to "Tournament". I agree that this is a small point - the only tangible real difference I see to having a prefix is autocompletion but the autompletion argument exists for the suffix approach as well.

:+1: for `axelrod_first_tournament_strategies` though good idea :+1:

I'm happy to start picking this up over next couple of weeks if no one else does. I believe there are essentially two steps (that don't necessarily need to be done in order):

1. Rename (according to whatever naming convention) the strategies from Axelrod's second tournament. (This should not be a complicated exercise but will require keeping track of things on `axelrod_fortran` in time for the next release - we should open an issue to keep track of the PRs and the renaming.)
2. Implement the strategies from the first.

**Comment by marcharper** [comment]
sgtm

**Comment by drvinceknight** [comment]
> Graaskamp round (1) is in the second tournament source code and claims to match the first round source.

@marcharper I'm not sure I see where you see that? Looking at https://github.com/Axelrod-Python/TourExec/blob/master/src/strategies/grasr.f I don't see it.

Looking at the Fortran and https://github.com/Axelrod-Python/Axelrod/pull/1115/files#diff-c699c4cf8dfbb23732ec03401295fedeR12 it doesn't match the explanation in the text but has some clear differences so unless I'm missing something I think Graaskamp modified the strategy for the second tournament. Thus, we should implement (for the first) based on description (clearly highlighting assumptions that need to be made) - and implement (for the second) based on the Fortran.

FYI this is the description in the paper which is currently what is implemented except for the single "and its own twin" part that we had missed.

> This rule plays tit for tat for 50 moves, defects on move 51, and then plays 5 more moves of tit for tat. A check is then made to see if the player seems to be RANDOM, in which case it defects from then on. A check is also made to see if the other is TIT FOR TAT, ANALOGY (a program from the preliminary tournament), and its own twin, in which case it plays tit for tat. Otherwise it randomly defects every 5 to 15 moves, hoping that enough trust has been built up so that the other player will not notice these defections.

**Comment by marcharper** [comment]
> @marcharper I'm not sure I see where you see that? Looking at https://github.com/Axelrod-Python/TourExec/blob/master/src/strategies/grasr.f I don't see it.

Earlier in the [original file](http://www-personal.umich.edu/%7Eaxe/research/Software/CC/CC2/TourExec1.1.f.html) it says

```
C TYPED BY AX, 3/27/79 (SAME AS ROUND ONE GRAASKAMP)
      INTEGER GRASR
c Time parameter elminated Ax 7/93
      K78R=GRASR(J,M,K,L,R,JA)
      RETURN
      END
```

(There are a few other lines that also say "same as round one ...")

If there's a mismatch with the description then we'll have to use our best judgement. It seems more likely to me that the description is incorrect than the comment (assuming it's a "copy/paste" situation) and less open to interpretation, but if it's very different...

**Comment by drvinceknight** [comment]
Right I see that now: thanks!

**Comment by drvinceknight** [comment]
There are 5 strategies with the "same as round one" annotation:

- Rapaport's TfT
- Nydegger
- Downing
- Graaskamp
- Smith's Tit For 2 Tats

*but* Smith's TF2T's wasn't in the first tournament so it is possible that "same as round one" doesn't mean what we think...

Just noting this for now (in case I'm missing something) - will carry on prepping this PR and we can discuss there when there's actual code to comment on.

**Comment by marcharper** [comment]
Interesting. For Downing it says "TYPED BY AX, 3/27/79 (SAME AS ROUND ONE REV.DOWNING)", but Downing wasn't revised until after the first tournament, so indeed it appears not to mean quite what we thought.

**Comment by drvinceknight** [comment]
I've finished working on a first attempt at this: https://github.com/Axelrod-Python/Axelrod/pull/1275

> For Downing it says "TYPED BY AX, 3/27/79 (SAME AS ROUND ONE REV.DOWNING)", but Downing wasn't revised until after the first tournament, so indeed it appears not to mean quite what we thought.


I think the "first tournament" might actually mean the first time the SECOND tournament was run with some of this Fortran code having been re-edited (?) in the 90s (I think)... So as you'll see on my PR I've only gone based on the description in Axelrod's paper and also the papers he cites there. I've added a lot to the docstrings in those cases.

**Comment by drvinceknight** [comment]
@id428 thanks again for bringing this to our attention, #1275 has been merged in to master and fixes the implementation errors. I'm going to try and put a new release together today (if I don't have time I'll get it done tomorrow).
